pkg/controller/vsphere/util.go

```diff
@@ -252,3 +253,49 @@ func ProviderStatusFromRawExtension(rawExtension *runtime.RawExtension) (*machin
 func isNotFoundErr(err error) bool {
 	return err != nil && strings.HasSuffix(err.Error(), http.StatusText(http.StatusNotFound))
 }
+
+// podPredicate is a predicate function for filtering PodList
+type podPredicate func(corev1.Pod) bool
+
+// isTerminating is a predicate for determine pods in 'Terminating' state
+func isTerminating(p corev1.Pod) bool {
+	return p.DeletionTimestamp != nil
+}
+
+// filterPods filters a podList and returns a PodList matching passed predicates
+func filterPods(podList *corev1.PodList, predicates ...podPredicate) *corev1.PodList {
+	filteredPods := &corev1.PodList{}
+	for _, pod := range podList.Items {
+		var match = true
+		for _, p := range predicates {
+			if !p(pod) {
+				match = false
+				break
+			}
+		}
+		if match {
+			filteredPods.Items = append(filteredPods.Items, pod)
+		}
+	}
+
+	return filteredPods
+}
+
+// getPodList returns pod list on a given node matching pod predicates
+func getPodList(ctx context.Context, apiReader runtimeclient.Reader, n *corev1.Node, filters []podPredicate) (*corev1.PodList, error) {
+
+	fieldSelector, err := fields.ParseSelector("spec.nodeName=" + n.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	allPods := &corev1.PodList{}
+	err = apiReader.List(ctx, allPods, &runtimeclient.ListOptions{
+		FieldSelector: fieldSelector,
+	})
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Nit, collapse this and don't shadow
```suggestion
	if err := apiReader.List(ctx, allPods, &runtimeclient.ListOptions{
		FieldSelector: fieldSelector,
	}); err != nil {
		return nil, err
	}
```